agent/engine/docker_task_engine_test.go

```diff
@@ -781,6 +769,242 @@ func TestTaskTransitionWhenStopContainerTimesout(t *testing.T) {
 	default:
 	}
 }
+
+// TestTaskTransitionWhenStopContainerReturnsUnknownBeforeSucceeding tests
+// if a task transitions to stopped only after StopContainer succeeds and
+// doesn't tranisition to stopped when the first StonContainer API returns an
+// unknown error.
+func TestTaskTransitionWhenStopContainerReturnsUnknownBeforeSucceeding(t *testing.T) {
+	ctrl, client, mockTime, taskEngine, _, imageManager := mocks(t, &defaultConfig)
+	defer ctrl.Finish()
+
+	sleepTask := testdata.LoadTask("sleep5")
+
+	eventStream := make(chan DockerContainerChangeEvent)
+
+	dockerEvent := func(status api.ContainerStatus) DockerContainerChangeEvent {
+		meta := DockerContainerMetadata{
+			DockerID: "containerId",
+		}
+		return DockerContainerChangeEvent{Status: status, DockerContainerMetadata: meta}
+	}
+
+	client.EXPECT().ContainerEvents(gomock.Any()).Return(eventStream, nil)
+	mockTime.EXPECT().After(gomock.Any()).AnyTimes()
+	containerStoppingError := DockerContainerMetadata{
+		Error: &CannotXContainerError{"Stop", "Error stopping container"},
+	}
+	for _, container := range sleepTask.Containers {
+		gomock.InOrder(
+			imageManager.EXPECT().AddAllImageStates(gomock.Any()).AnyTimes(),
+			client.EXPECT().PullImage(container.Image, nil).Return(DockerContainerMetadata{}),
+			imageManager.EXPECT().AddContainerReferenceToImageState(container),
+			imageManager.EXPECT().GetImageStateFromImageName(gomock.Any()).Return(nil),
+			client.EXPECT().CreateContainer(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Do(
+				func(x, y, z, timeout interface{}) {
+					go func() { eventStream <- dockerEvent(api.ContainerCreated) }()
+				}).Return(DockerContainerMetadata{DockerID: "containerId"}),
+			// StartContainer returns timeout error. This should cause the engine
+			// to transition the task to STOPPED and to stop all containers of
+			// the task
+			client.EXPECT().StartContainer("containerId", startContainerTimeout).Return(DockerContainerMetadata{
+				Error: &DockerTimeoutError{},
+			}),
+			// StopContainer returns an error that we don't know about
+			client.EXPECT().StopContainer("containerId", gomock.Any()).Return(containerStoppingError),
+			// Since task is not in steady state, progressContainers causes
+			// another invocation of StopContainer. Return an
+			// unretriable error for this, which should cause the task engine
```

Comment:
Nope. Instead of `return an unretriable error`, it should read `return success` or something. I'll modify this.